[Bug]: make_norm_from_scale should create picklable classes even when used in-line.

### Bug summary

The new `matplotlib.colors.make_norm_from_scale` helper dynamically generates a norm class from a scale class.  Currently, in the codebase, it is only used as a decorator to create "toplevel" classes (e.g., it is used to generate LogNorm from LogScale, etc.), but it can also be used within other functions to dynamically generate a norm class based on a user-given arbitrary scale (see #20752 for an example of application, which is however not necessary to understand for what follows).  In the latter case, the dynamically generate class is currently not picklable (because pickling of classes relies on the existence of global names, see e.g. ["classes are pickled by named reference"](https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled )).  It would be generally useful to get rid of this restriction, which can be done by implementing the [`__reduce__` protocol](https://docs.python.org/3/library/pickle.html#object.__reduce__); there's already other examples in the codebase of dynamically generated classes that use the same mechanism).

I'm tagging this as "good first issue" because there's no API design and the eng goal is clear, but medium (perhaps hard) difficulty because it requires somewhat sophisticated understanding of the details of the pickling process.

### Code for reproduction

```python
pickle.dumps(matplotlib.colors.make_norm_from_scale(matplotlib.scale.LogitScale, matplotlib.colors.Normalize))
```


### Actual outcome

```
Can't pickle <class 'matplotlib.colors.Normalize'>: it's not the same object as matplotlib.colors.Normalize
```
(Note the additional confusion here: there's two classes that are both named `matplotlib.colors.Normalize` -- the original one and the dynamically generated one -- but they are different.)

### Expected outcome

A correct round-trippable pickle.

### Operating system

_No response_

### Matplotlib Version

master (unreleased, pre 3.5)

### Matplotlib Backend

_No response_

### Python version

_No response_

### Jupyter version

_No response_

### Other libraries

_No response_

### Installation

source

### Conda channel

_No response_

I think this wasn't fully solved for the built-in classes of `matplotlib.colors`. Adding this test to `test_dynamic_norm`:

```python
    # Also test the builtin classes
    norm = mpl.colors.LogNorm(vmin=0.01, vmax=100)

    assert type(pickle.loads(pickle.dumps(norm))) \
        == type(norm)
```

Fails with:

```
E       _pickle.PicklingError: Can't pickle <class 'matplotlib.colors.LogNorm'>: it's not the same object as matplotlib.colors.LogNorm
```

--- a/lib/matplotlib/colors.py
+++ b/lib/matplotlib/colors.py
@@ -43,6 +43,7 @@ import base64
 from collections.abc import Sized, Sequence
 import copy
 import functools
+import importlib
 import inspect
 import io
 import itertools
@@ -1528,9 +1529,22 @@ def _make_norm_from_scale(scale_cls, base_norm_cls, bound_init_signature):
 
     class Norm(base_norm_cls):
         def __reduce__(self):
+            cls = type(self)
+            # If the class is toplevel-accessible, it is possible to directly
+            # pickle it "by name".  This is required to support norm classes
+            # defined at a module's toplevel, as the inner base_norm_cls is
+            # otherwise unpicklable (as it gets shadowed by the generated norm
+            # class).  If either import or attribute access fails, fall back to
+            # the general path.
+            try:
+                if cls is getattr(importlib.import_module(cls.__module__),
+                                  cls.__qualname__):
+                    return (_create_empty_object_of_class, (cls,), vars(self))
+            except (ImportError, AttributeError):
+                pass
             return (_picklable_norm_constructor,
                     (scale_cls, base_norm_cls, bound_init_signature),
-                    self.__dict__)
+                    vars(self))
 
         def __init__(self, *args, **kwargs):
             ba = bound_init_signature.bind(*args, **kwargs)
@@ -1603,9 +1617,12 @@ def _make_norm_from_scale(scale_cls, base_norm_cls, bound_init_signature):
     return Norm
 
 
-def _picklable_norm_constructor(*args):
-    cls = _make_norm_from_scale(*args)
+def _create_empty_object_of_class(cls):
     return cls.__new__(cls)
+
+
+def _picklable_norm_constructor(*args):
+    return _create_empty_object_of_class(_make_norm_from_scale(*args))
 
 
 @make_norm_from_scale(

--- a/lib/matplotlib/tests/test_pickle.py
+++ b/lib/matplotlib/tests/test_pickle.py
@@ -221,6 +221,11 @@ def test_mpl_toolkits():
     assert type(pickle.loads(pickle.dumps(ax))) == parasite_axes.HostAxes
 
 
+def test_standard_norm():
+    assert type(pickle.loads(pickle.dumps(mpl.colors.LogNorm()))) \
+        == mpl.colors.LogNorm
+
+
 def test_dynamic_norm():
     logit_norm_instance = mpl.colors.make_norm_from_scale(
         mpl.scale.LogitScale, mpl.colors.Normalize)()

